File: frontend/lib/schemas/patientStatus.ts
/**
 * 患者ステータス連動 (docs/plans/patient-status-schedule-design-2026-09-09.md §7-3)。
 * BE `backend/app/schemas/patient_status.py` と 1:1 (キー名を変えない)。
 *
 * - GET  /api/v1/patients/{id}/status-impact?to=&from_date=  → statusImpactSchema
 * - POST /api/v1/patients/{id}/status-change                 → statusChangeRequestSchema / statusChangeResultSchema
 */
import { z } from 'zod';

import { statusEnum } from './patient';
import { patientV2ReadSchema } from './v2/patient';

export const statusDirectionEnum = z.enum(['deactivate', 'reactivate', 'none']);
export const specialPeriodActionEnum = z.enum(['keep', 'end']);

export const weekCountSchema = z.object({
  iso_year: z.number().int(),
  iso_week: z.number().int(),
  count: z.number().int().nonnegative(),
  label: z.string(),
});

export const impactVisitsSchema = z.object({
  total: z.number().int().nonnegative(),
  by_week: z.array(weekCountSchema).default([]),
  by_source: z.record(z.number().int()).default({}),
  pair_groups: z.number().int().nonnegative().default(0),
  excluded: z.record(z.number().int()).default({}),
});

export const impactSpecialPeriodSchema = z.object({
  id: z.string().uuid(),
  start_date: z.string(),
  end_date: z.string(),
  pool_marks: z.number().int().nonnegative(),
  placed_marks: z.number().int().nonnegative(),
  placed_future_visits: z.number().int().nonnegative(),
});

export const impactRegenerateSchema = z.object({
  weeks: z.array(weekCountSchema).default([]),
  total: z.number().int().nonnegative().default(0),
});

export const statusImpactSchema = z.object({
  patient_id: z.string().uuid(),
  current_status: z.string(),
  to_status: statusEnum,
  from_date: z.string(),
  direction: statusDirectionEnum,
  visits: impactVisitsSchema,
  special_period: impactSpecialPeriodSchema.nullish(),
  fixed_visit_rows: z.number().int().nonnegative().default(0),
  pending_requests: z.number().int().nonnegative().default(0),
  kaipoke_weeks: z.number().int().nonnegative().default(0),
  regenerate: impactRegenerateSchema.nullish(),
});
export type StatusImpact = z.infer<typeof statusImpactSchema>;

export const statusChangeRequestSchema = z.object({
  status: statusEnum,
  from_date: z.string().optional(),
  special_period_action: specialPeriodActionEnum.default('keep'),
  regenerate: z.boolean().default(true),
  note: z.string().max(200).optional(),
});
export type StatusChangeRequest = z.input<typeof statusChangeRequestSchema>;

export const opGroupRefSchema = z.object({
  iso_year: z.number().int(),
  iso_week: z.number().int(),
  op_group_id: z.string().uuid(),
});

export const statusChangeResultSchema = z.object({
  // BE は PatientV2Read (null を含む) を返す。フォーム用 patientReadSchema は
  // sex_restriction/note の null を受けないため、2026-09-10 に本番で
  // 「稼働中に戻す」が ZodError で失敗した (BE は 200 で完了済み)。読取専用の
  // v2 スキーマで受け、表示に不要な差異でユーザー操作を失敗扱いにしない。
  patient: patientV2ReadSchema.extend({
    status_changed_at: z.string().nullish(),
    status_changed_by: z.string().uuid().nullish(),
  }),
  direction: statusDirectionEnum,
  cancelled_visit_ids: z.array(z.string().uuid()).default([]),
  cancelled_count: z.number().int().nonnegative().default(0),
  special_period: z
    .object({
      id: z.string().uuid(),
      action: specialPeriodActionEnum,
      cancelled_pool_marks: z.number().int().nonnegative().default(0),
    })
    .nullish(),
  rejected_requests: z.number().int().nonnegative().default(0),
  op_groups: z.array(opGroupRefSchema).default([]),
  regenerated: z
    .object({
      created: z.number().int().nonnegative(),
      weeks: z.array(weekCountSchema).default([]),
    })
    .nullish(),
  notification_count: z.number().int().nonnegative().default(0),
});
export type StatusChangeResult = z.infer<typeof statusChangeResultSchema>;

/**
 * 422 `{"detail": {"code": "patient_not_active", ...}}` (Phase 2 入口ガード・§7-3 (d))。
 * `can_override` が true なら「稼働中にして続ける」導線を出す。
 */
export const patientNotActiveDetailSchema = z.object({
  code: z.literal('patient_not_active'),
  patient_id: z.string().uuid(),
  status: z.string(),
  status_label: z.string(),
  can_override: z.boolean().default(false),
  message: z.string(),
});
export type PatientNotActiveDetail = z.infer<typeof patientNotActiveDetailSchema>;

File: backend/app/schemas/patient_status.py
"""患者ステータス連動 (docs/plans/patient-status-schedule-design-2026-09-09.md §7-3) の pydantic.

- ``GET  /api/v1/patients/{id}/status-impact``  → :class:`StatusImpact`
- ``POST /api/v1/patients/{id}/status-change``  → :class:`StatusChangeRequest` / :class:`StatusChangeResult`

契約は設計書 §7-3 (a)(b) を正とし、この module の型がそのまま FE の
``frontend/lib/schemas/patientStatus.ts`` と対になる (キー名を変えない)。
"""

from __future__ import annotations

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.v2.patient import PatientStatusV2, PatientV2Read

StatusDirection = Literal["deactivate", "reactivate", "none"]
SpecialPeriodAction = Literal["keep", "end"]


class WeekCount(BaseModel):
    """ISO 週ごとの件数 (表示用ラベル付き)."""

    model_config = ConfigDict(extra="forbid")

    iso_year: int
    iso_week: int
    count: int = Field(ge=0)
    label: str = Field(description="例: '9/14週' (週の月曜日)")


class ImpactVisits(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total: int = Field(ge=0, description="取消対象 (planned・from_date 以降・除外分を引いた数)")
    by_week: list[WeekCount] = []
    by_source: dict[str, int] = Field(default_factory=dict)
    pair_groups: int = Field(ge=0, default=0, description="2 名体制 (visit_group_id) のグループ数")
    excluded: dict[str, int] = Field(
        default_factory=dict,
        description="対象外にした理由別件数: checked_in / in_progress / week_pinned",
    )


class ImpactSpecialPeriod(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    start_date: date
    end_date: date
    pool_marks: int = Field(ge=0)
    placed_marks: int = Field(ge=0)
    placed_future_visits: int = Field(
        ge=0, description="配置済み ● のうち from_date 以降の planned"
    )


class ImpactRegenerate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    weeks: list[WeekCount] = Field(
        default_factory=list, description="expected = 型から作られる見込み件数"
    )
    total: int = Field(ge=0, default=0)


class StatusImpact(BaseModel):
    """``GET /patients/{id}/status-impact`` の応答."""

    model_config = ConfigDict(extra="forbid")

    patient_id: UUID
    current_status: str
    to_status: PatientStatusV2
    from_date: date
    direction: StatusDirection
    visits: ImpactVisits
    special_period: ImpactSpecialPeriod | None = None
    fixed_visit_rows: int = Field(ge=0, default=0)
    pending_requests: int = Field(ge=0, default=0)
    kaipoke_weeks: int = Field(ge=0, default=0, description="送信差分が出る見込みの週数")
    regenerate: ImpactRegenerate | None = None


class StatusChangeRequest(BaseModel):
    """``POST /patients/{id}/status-change`` の要求."""

    model_config = ConfigDict(extra="forbid")

    status: PatientStatusV2
    from_date: date | None = Field(
        default=None, description="取消/再生成の起点 (JST)。省略時は今日。過去日は 422"
    )
    special_period_action: SpecialPeriodAction = Field(
        default="keep",
        description="deactivate 時のみ有効。keep=特別訪問週間に触らない (既定・PO 決定) / end=終了して ○ を取消",
    )
    regenerate: bool = Field(
        default=True, description="reactivate 時のみ有効。型から生成済み週に予定を作り直す"
    )
    note: str | None = Field(default=None, max_length=200)


class OpGroupRef(BaseModel):
    model_config = ConfigDict(extra="forbid")

    iso_year: int
    iso_week: int
    op_group_id: UUID


class RegeneratedSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    created: int = Field(ge=0)
    weeks: list[WeekCount] = Field(default_factory=list)


class SpecialPeriodResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    action: SpecialPeriodAction
    cancelled_pool_marks: int = Field(ge=0, default=0)


class StatusChangeResult(BaseModel):
    """``POST /patients/{id}/status-change`` の応答."""

    model_config = ConfigDict(extra="forbid")

    patient: PatientV2Read
    direction: StatusDirection
    cancelled_visit_ids: list[UUID] = Field(default_factory=list)
    cancelled_count: int = Field(ge=0, default=0)
    special_period: SpecialPeriodResult | None = None
    rejected_requests: int = Field(ge=0, default=0)
    op_groups: list[OpGroupRef] = Field(default_factory=list)
    regenerated: RegeneratedSummary | None = None
    notification_count: int = Field(ge=0, default=0)
